atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionManager.java

```diff
@@ -274,17 +273,18 @@ public <T, E extends Exception> T finishWithCallback(TransactionTask<T, E> task,
 
             TransactionTask<T, E> wrappedTask = wrapTaskIfNecessary(task, immutableTsLock);
 
-            CallbackAwareTransaction tx = delegate;
+            CallbackAwareTransaction txn = delegate;
             T result;
             try {
-                result = runTaskThrowOnConflictWithCallback(wrappedTask, tx, callback);
+                txn.onCommitOrAbort(txn::reportExpectationsCollectedData);
+                txn.onCommitOrAbort(callback);
+                result = runTaskThrowOnConflict(wrappedTask, txn);
```

Comment:
Now this reads funny with this diff, because the transaction is not in a try-with-resources or anything.
But runTaskThrowOnConflict HAS TO either commit or abort and therefore run the cleanup.

Reply:
It would MAYBE read a bit better if we pushed the txn.onCommitOrAbort calls into the wrapped task (since technically we have some code that's not wrapped in cleanup logic)

Reply:
I think explicitly saying `onCommitOrAbort` is not allowed to throw is fine (and we probably want that?)

Reply:
I am certainly happy with this for now.